### PR TITLE
SNS - Interface

### PR DESCRIPTION
**Pull request by SnickeyX**
Goal of the I2c sensor interface is to be able to use this abstract class with the code for the relevant mux. Having this interface allows us to generalize the mux code to handle any I2C sensor connected to it. 

- [x] Agree on the best return type for the read function.

### CODE REVIEW

**Comment by SnickeyX** [comment]
The biggest issue for me is setting the correct return type for read - I have temporarily kept it at uint16_t which works with current I2C sensors that I am aware of but this is bound to change.

File: lib/sensors/sensors.hpp
Comment (kshxtij) [inline lib/sensors/sensors.hpp:20]: https://github.com/Hyp-ed/hyped-2023/blob/9a3285262877e8cb0e3a3bdb00dadacc28cbcf76/lib/io/i2c.cpp#L43

Would we want the sensor returning an optional since the i2c interface does?

Comment (SnickeyX) [reply lib/sensors/sensors.hpp:20]: Good point, will change

File: lib/sensors/sensors.hpp
Comment (mifrandir) [inline lib/sensors/sensors.hpp:21]: Do we need both read and write access at all times? Surely most users will only require one of these.

**Review by mifrandir — CHANGES_REQUESTED** [review]
I perfer the design where we have one toplevel instance/manager/factory (`II2c`) that allows us to create sensors. With this design you need to make all the sensors on startup and pass them in to the main function (rather than providing callees with the means to read to/write from I2c at will). That's what I did for GPIO. Is there any reason why this wouldn't work here? Besides not being excessively verbose, this has the neat effect that we can manage file descriptors within this instance and don't need the callees to be efficient.

File: lib/sensors/sensors.hpp
Comment (SnickeyX) [reply lib/sensors/sensors.hpp:21]: On second thought, the write function is redundant since thats what configure essentially does. Will remove that. We definitely need read though.

Comment (mifrandir) [reply lib/sensors/sensors.hpp:21]: Maybe call it `I2cReader` then? For consistency.

Comment (SnickeyX) [reply lib/sensors/sensors.hpp:21]: But the classes deriving this will be like `ITemperature` or `IAccelerometer`  which won't just be readers?

Comment (kshxtij) [reply lib/sensors/sensors.hpp:18]: FWIW, we need a generic II2cSensor interface specifically because of the Multiplexer class that needs to be implemented soon. The Multiplexer class would need to accept II2c instances as a hardware mux could have any type of i2c sensor connected to it.

Comment (mifrandir) [reply lib/sensors/sensors.hpp:18]: > But the classes deriving this will be like `ITemperature` or `IAccelerometer` which won't just be readers?

I don't see anything wrong with this. Also, I don't think these will derive from this but rather wrap around it. I.e. they add some sort of code that interprets the `std::optional<uint16_t>` as a temperature or an acceleration and possibly does some setup.

If you really want to make these abstract (`ITemperature` vs `Temperature`) and it wasn't just a typo I don't think an accelerometer interface should have anything to do with I2C specifically.

> FWIW, we need a generic II2cSensor interface specifically because of the Multiplexer class that needs to be implemented soon. The Multiplexer class would need to accept II2c instances as a hardware mux could have any type of i2c sensor connected to it.

Let's chat tonight? Not sure I agree or understand.

File: lib/sensors/main.cpp
Comment (kshxtij) [inline lib/sensors/main.cpp:1]: just leave this file empty, we need a cpp file in the directory but don't need it to have anything. 

This file should be deleted in one of the temperature or accelerometer prs anyways as soon as we have an actual cpp file in here

**Review by kshxtij — CHANGES_REQUESTED** [review]
LGTM, just clear that file imo

File: lib/sensors/main.cpp
Comment (mifrandir) [reply lib/sensors/main.cpp:1]: there shouldn't be anything called `main.cpp` or `main.hpp` imo

Comment (mifrandir) [reply lib/sensors/main.cpp:1]: Maybe an empty `sensors.cpp`?

**Review by mifrandir — APPROVED** [review]

File: lib/sensors/main.cpp
Comment (SnickeyX) [reply lib/sensors/main.cpp:1]: yeah will delete after either temp or acc get approved and merged into here

Comment (kshxtij) [reply lib/sensors/main.cpp:1]: @SnickeyX we don't need this pr to be open until the others are done. Replace main.cpp with an empty sensors.cpp, we merge this pr and then the other prs automatically shift bases to master.

Comment (SnickeyX) [reply lib/sensors/main.cpp:1]: W

**Review by kshxtij — APPROVED** [review]

**Pull request merged**
